File: shared/Aegis/aegis_ts/src/firebase-token-verifier.ts
import admin from 'firebase-admin';
import type { ITokenVerifier, DecodedToken } from './interfaces/token-verifier.js';

/**
 * Options for initializing the {@link FirebaseTokenVerifier}.
 *
 * These are intentionally Firebase-agnostic names so that consuming services
 * never need to import or reference `firebase-admin` directly.
 */
export interface FirebaseAuthOptions {
  /** Firebase / Google Cloud project identifier. */
  readonly projectId: string;
  /**
   * Service-account credentials as a raw JSON string **or** a Base64-encoded
   * JSON string.  When omitted, Application Default Credentials (ADC) are used.
   */
  readonly serviceAccountKey?: string;
}

/**
 * Firebase implementation of {@link ITokenVerifier}.
 *
 * Verifies Firebase ID tokens using the Firebase Admin SDK.  The underlying
 * Firebase app is initialised lazily on construction: if an app has already
 * been initialised (e.g. by Synapse), it is reused rather than creating a
 * second instance.
 *
 * @example
 * ```typescript
 * import { FirebaseTokenVerifier } from '@elastic-resume-base/aegis';
 *
 * const verifier = new FirebaseTokenVerifier({
 *   projectId: process.env.FIREBASE_PROJECT_ID ?? 'demo-project',
 * });
 *
 * const decoded = await verifier.verifyToken(idToken);
 * console.log(decoded.uid, decoded.email);
 * ```
 */
export class FirebaseTokenVerifier implements ITokenVerifier {
  private readonly app: admin.app.App;

  constructor(options: FirebaseAuthOptions) {
    if (admin.apps.length > 0) {
      this.app = admin.apps[0]!;
    } else {
      const appOptions: admin.AppOptions = { projectId: options.projectId };

      if (options.serviceAccountKey) {
        try {
          const raw = options.serviceAccountKey.trim();
          const decoded = raw.startsWith('{')
            ? raw
            : Buffer.from(raw, 'base64').toString('utf-8');
          const credentials = JSON.parse(decoded) as admin.ServiceAccount;
          appOptions.credential = admin.credential.cert(credentials);
        } catch {
          // Parsing failed — fall back to Application Default Credentials.
        }
      }

      this.app = admin.initializeApp(appOptions);
    }
  }

  /**
   * Verifies a Firebase ID token and returns the decoded claims.
   *
   * @param token - The raw Firebase ID token string from the client.
   * @returns Decoded token data.
   * @throws If the token is invalid or expired.
   */
  async verifyToken(token: string): Promise<DecodedToken> {
    const decoded = await admin.auth(this.app).verifyIdToken(token);
    return {
      uid: decoded.uid,
      email: decoded.email,
      name: decoded.name as string | undefined,
      picture: decoded.picture,
    };
  }
}

File: shared/Aegis/v2/aegis_py/aegis_py/firebase_token_verifier.py
"""Firebase Admin SDK token verifier implementation.

Wraps ``firebase_admin.auth.verify_id_token`` in the :class:`ITokenVerifier`
interface so Aegis Python consumers are never directly coupled to
``firebase-admin``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import firebase_admin
from firebase_admin import auth as firebase_auth


@dataclass
class DecodedFirebaseToken:
    """Decoded Firebase ID token data exposed as typed attributes."""

    uid: str
    email: str | None
    name: str | None
    picture: str | None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "DecodedFirebaseToken":
        return cls(
            uid=str(data["uid"]),
            email=data.get("email"),
            name=data.get("name"),
            picture=data.get("picture"),
        )


class FirebaseTokenVerifier:
    """Token verifier backed by the Firebase Admin SDK.

    This class must be used **after** :func:`~aegis_py.initialize_auth` has
    been called to initialise the Firebase Admin SDK.

    Args:
        app: Optional ``firebase_admin.App`` instance.  Uses the default app
            when *app* is ``None``.
    """

    def __init__(self, app: firebase_admin.App | None = None) -> None:
        self._app = app

    async def verify_token(self, token: str) -> DecodedFirebaseToken:
        """Verify a Firebase ID token and return the decoded token data.

        Args:
            token: Raw JWT bearer token (without the ``Bearer `` prefix).

        Returns:
            :class:`DecodedFirebaseToken` with ``uid``, ``email``, ``name``,
            and ``picture`` fields.

        Raises:
            ValueError: If the token is invalid, expired, or revoked.
        """
        try:
            decoded: dict[str, Any] = firebase_auth.verify_id_token(
                token, app=self._app
            )
        except firebase_auth.RevokedIdTokenError as exc:
            raise ValueError(f"Revoked token: {exc}") from exc
        except firebase_auth.UserDisabledError as exc:
            raise ValueError(f"User disabled: {exc}") from exc
        except firebase_auth.ExpiredIdTokenError as exc:
            raise ValueError(f"Expired token: {exc}") from exc
        except firebase_auth.InvalidIdTokenError as exc:
            raise ValueError(f"Invalid token: {exc}") from exc
        except Exception as exc:
            raise ValueError(f"Token verification failed: {exc}") from exc

        return DecodedFirebaseToken.from_dict(decoded)
